### PR TITLE
Partial network rewrite

### PR DESCRIPTION
**Pull request by S-S-X**
## TESTING REQUIRED

ANY OLD BUGS THAT ALSO APPEAR BEFORE THIS PR SHOULD NOT BE FIXED IN THIS BRANCH

Found new bugs so far:
- [x] Supply converters are not working correctly, this seems to be because of machine timeout ABM change.
       actually they do seem to work but keeps blinking infotext between "no network" and actual supply info.
       Tests added :heavy_check_mark:
- [x] Placing cable between other same tier cables and machine wont add machine to network (but other cables without network and machines after those will be added) Digging and placing machine again fixes connection.
       Tests added :heavy_check_mark:
- [x] Digging/replacing connecting intermediate cable and then digging cable below machine wont remove machine from network, machine works and is part of network also drawing power normally from network.
       Tests added :heavy_check_mark:
- [x] Digging machine will remove neighbor machines from network (but wont touch cables).
       Digging machines should not affect other machines at all, machine here behaves like cable and it should not.
       Tests added :heavy_check_mark:
- [x] Placing machine adds neighbor unconnected cables to same network machine is added to.
       While placing machine should not affect unconnected cables it should still check for cables connected to another network and cause overload immediately, better sooner than later.
      This invalid network part is lost when cache is built again, only affects placement. Can be annoying for sure if you don't know how it behaves.
       Tests added :heavy_check_mark:
- [x] Digging single cable will remove machine from network even if machine has multiple connected cables to same network
      Tests added :heavy_check_mark:
- [x] Placing new cables around machine will always add machines to network, this can cause infinite duplicates in network.
      Tests added :heavy_check_mark:
- [x] Machines wont time out ever when removed from network by digging cables, this causes infotext "has no network" to be immediately replaced with "idle" or "unpowered".
      No tests added :x: 
- [x] Networks wont overload immediately after placing machine between 2 different networks, not actually big problem but nice to have also for future updates (less aggressive overloads). This kind of worked before as caches were built a lot more often.
      Tests added :heavy_check_mark:
- [x] Supply converter inserts duplicates to RE_nodes when cable connecting to network is placed nearby.
      No tests added :x: 
- [x] When supply converter is placed between clean networks it does not attach to networks correctly. This probably has something to do with cable tiers as when placed machines only look for first found compatible tier for cables.
      Tests added :heavy_check_mark:
- [x] Digging supply converter does not correctly remove it from all networks, whole multi tier machine support is broken.
      Tests added :heavy_check_mark:

Just for the record actual changed LoC is not as bad as it looks like. Excluding test related stuff which causes 16 files changed, 2434 insertions(+), actual code changes are:
9 files changed, 982 insertions(+), 745 deletions(-)
Might be possible to optimize that bit but not sure if it is worth it here, there's other upcoming updates planned anyway.

https://github.com/mt-mods/technic/pull/96#issuecomment-714789819

### Partial power network rewrite. Testing required, WIP, back off if you fear bugs
Opening PR for partially rewritten power network handler. Breaks compatibility if some other mod depends on SS metadata, however for special cases where it cannot be easily fixed on other mod it is very easy to add metadata back and update it with ABM.

#### Performance
Performance seems to be slightly better that what it was, with worst case network (built badly, multiple switching stations etc.) performance boost during testing (averages for more than 8k measurements) was over 50%. With "normal" networks not that much but still better. Didn't find any case where performance would be worse than before.

#### Help wanted
Most probably there will be bugs, for that reason I added label "Help wanted" hoping that someone would have time to test things and write/post screenshots about problems found. maybe even fix problems.

#### Highlights / most important stuff
Most important improvements:
* Switching station is not anymore network controller, actually it is not even registered as technic machine.
  It is still needed as it does add heart beat signal for network (resets network TTL countdown).
* Every switching station now displays supply / demand just like power monitor.
* Second ABM for switching stations removed.
* Globalstep executes networks directly instead of through switching station.
* Cache rebuilding is not triggered as often as it was before.
* Cache building is cheaper, flattening arrays removed by writing  indexed arrays from beginning.
* More reusable code and is actually reused.
* Digging and building network nodes should be bit better and cheaper but no hard proofs on this.
* Added unit testing with busted along with some tests.

Removed metadata from switching station (moved values to network hash):
* active
* active_pos
* supply
* demand
* battery_count
* battery_charge
* battery_charge_max

### Metadata cleanup

Added LBM (hardcoded disable) to remove metadata from technic machines, contains metadata keys for this PR and previous merged PR's:
`"LV_EU_timeout", "MV_EU_timeout", "HV_EU_timeout", "LV_network", "MV_network", "HV_network", "active_pos", "supply", "demand", "battery_count", "battery_charge", "battery_charge_max", "active"`

### CODE REVIEW

**Comment by S-S-X** [comment]
**Networks should be fairly stable**. *Should*... I've not tested anything but very simple things for development, while it should be stable it is very well possible that it is not stable at all.
**Network building still work in progress**, this is still mostly old stuff. Should work similar way like it did before, still good to test as there's also been some changes.

Things to test, good to have hash for commit that was used during testing:
* Networks working correctly.
* Amount of switching stations is unlimited, no switching station fights at any level, switching stations wont affect network performance.
* Amount of power monitors is unlimited, power monitors wont affect network performance.
* Network collection for globalstep and execution through globalstep works correctly.
* Network caches work correctly: no unnecessary cleanups, rebuilding all the time without reason etc.
* Other mods that can use technic networks works correctly:
  * Jumpdrive (HV connector)
  * Digtron (engine)
  * ...?
* Many other things have also changed... just test whatever.
* Always good to check whatever general random stuff.

**Comment by S-S-X** [comment]
@tuedel reported:
[21:20.16] tuedel i found a bug in your network-ng branch, but i'm not sure if this part is still work in progress
[21:20.17] SX wip part is mostly in register/cables.lua
[21:21.10] tuedel ok, so i'm guessing this is a known bug then
[21:21.11] SX that one is actually bug caused by switching network processing to indexed arrays
[21:21.45] SX no it was not actually known but I immediately just knew what causes it
[21:23.24] SX I was thinking about fixing that part of network building so that it wont visit same nodes multiple times but 
did not actually think about that bug, so noticing that increases priority for that stu

image

This has to be fixed and should get high priority. To fix simply check if node is already added or not, all_nodes is hash and should allow easy check for this.

**Comment by S-S-X** [comment]
Things noted:
`technic.cables` is hash table containing all cables and all machines. This has not changed.
`technic.networks[id].all_nodes` was hash table containing network cables but not machines. This is updated to contain also machines.

At least naming has been wrong, added FIXME note to unit tests about `technic.cables`, it seems that many pieces assumes that machines are there so machines cannot be just simply dropped from `technic.cables`.

**Comment by tuedel** [comment]
Experienced two crashes with 0d0a0bb:

While a player was connecting two networks (built before switching to this branch) with some machines but no switching station (I think):
```
2020-10-05 22:45:16: ERROR[Main]: ServerError: AsyncErr: ServerThread::run Lua: Runtime error from mod 'technic' in callback item_OnPlace(): Runtime error from mod 'technic' in callback node_on_construct(): ...netest/mods/technic/technic/machines/register/cables.lua:58: attempt to index local 'network' (a nil value)
2020-10-05 22:45:16: ERROR[Main]: stack traceback:
2020-10-05 22:45:16: ERROR[Main]:         ...netest/mods/technic/technic/machines/register/cables.lua:58: in function 'place_network_node'
2020-10-05 22:45:16: ERROR[Main]:         ...netest/mods/technic/technic/machines/register/cables.lua:182: in function <...netest/mods/technic/technic/machines/register/cables.lua:182>
2020-10-05 22:45:16: ERROR[Main]:         [C]: in function 'f'
2020-10-05 22:45:16: ERROR[Main]:         ...netest/.minetest/mods/monitoring/metrictypes/counter.lua:42: in function 'add_node'
2020-10-05 22:45:16: ERROR[Main]:         ...icia-minetest-5.3.0/share/minetest/builtin/game/item.lua:368: in function <...icia-minetest-5.3.0/share/minetest/builtin/game/item.lua:258>
2020-10-05 22:45:16: ERROR[Main]: stack traceback:
2020-10-05 22:45:16: ERROR[Main]:         [C]: in function 'f'
2020-10-05 22:45:16: ERROR[Main]:         ...netest/.minetest/mods/monitoring/metrictypes/counter.lua:42: in function 'add_node'
2020-10-05 22:45:16: ERROR[Main]:         ...icia-minetest-5.3.0/share/minetest/builtin/game/item.lua:368: in function <...icia-minetest-5.3.0/share/minetest/builtin/game/item.lua:258>
```

While a player was using a drill to dig a quarry:
```
2020-10-05 22:54:09: ERROR[Main]: ServerError: AsyncErr: ServerThread::run Lua: Runtime error from mod 'technic' in callback item_OnUse(): ...test/.minetest/mods/technic/technic/machines/network.lua:65: attempt to call field 'remove' (a nil value)
2020-10-05 22:54:09: ERROR[Main]: stack traceback:
2020-10-05 22:54:09: ERROR[Main]:         ...test/.minetest/mods/technic/technic/machines/network.lua:65: in function 'remove_network_node'
2020-10-05 22:54:09: ERROR[Main]:         ...netest/mods/technic/technic/machines/register/cables.lua:113: in function 'func'
2020-10-05 22:54:09: ERROR[Main]:         ....3.0/share/minetest/builtin/profiler/instrumentation.lua:106: in function 'callback'
2020-10-05 22:54:09: ERROR[Main]:         ...icia-minetest-5.3.0/share/minetest/builtin/game/item.lua:649: in function 'on_dig'
2020-10-05 22:54:09: ERROR[Main]:         ...st/.minetest/mods/technic/technic/tools/mining_drill.lua:62: in function 'drill_dig_it0'
2020-10-05 22:54:09: ERROR[Main]:         ...st/.minetest/mods/technic/technic/tools/mining_drill.lua:157: in function 'drill_dig_it'
2020-10-05 22:54:09: ERROR[Main]:         ...st/.minetest/mods/technic/technic/tools/mining_drill.lua:324: in function 'mining_drill_mk3_handler'
2020-10-05 22:54:09: ERROR[Main]:         ...st/.minetest/mods/technic/technic/tools/mining_drill.lua:416: in function <...st/.minetest/mods/technic/technic/tools/mining_drill.lua:415>
```

File: technic/machines/network.lua [hunk 1]
```diff
@@ -0,0 +1,587 @@
...
+function technic.remove_network_node(network_id, pos)
+	local network = technic.networks[network_id]
+	if not network then return end
+	-- Clear hash tables, cannot use table.remove
+	local node_id = poshash(pos)
+	technic.cables[node_id] = nil
+	network.all_nodes[node_id] = nil
+	-- Clear indexed arrays, do NOT leave holes
+	for _,tblname in ipairs(network_node_arrays) do
+		local table = network[tblname]
```
Comment (tuedel) [inline technic/machines/network.lua:61]: You should probably use some other name here :wink:

Comment (S-S-X) [reply technic/machines/network.lua:61]: probably, did not change names when I copied that loop from register/cables.lua
however my imagination is not any better than op's :smile:

Comment (S-S-X) [reply technic/machines/network.lua:61]: > Experienced two crashes with [0d0a0bb](https://github.com/mt-mods/technic/commit/0d0a0bbbb428563117fc748ce4eabbd2252c01f8):
> 
> While a player was connecting two networks (built before switching to this branch) with some machines but no switching station (I think):

Almost sure bug that caused crash was introduced in last significant commit which was dd8fcb4
That was not so well tested, especially digging not tested at all.
Also after breaking register/cables.lua apart I had similar crashes with placement but fixed most of those if not all, digging part was left without much love and apparently needs some work.

I'll see if I can reproduce it and possibly add unit test for situation.

**Comment by S-S-X** [comment]
> Experienced two crashes with [0d0a0bb](https://github.com/mt-mods/technic/commit/0d0a0bbbb428563117fc748ce4eabbd2252c01f8):
> 
> While a player was connecting two networks (built before switching to this branch) with some machines but no switching station (I think):

Almost sure bug that caused crash was introduced in last significant commit which was dd8fcb4
That was not so well tested, especially digging not tested at all.
Also after breaking register/cables.lua apart I had similar crashes with placement but fixed most of those if not all, digging part was left without much love and apparently needs some work.

I'll see if I can reproduce it and possibly add unit test for situation.

**Comment by S-S-X** [comment]
Well these both were just simply "not using brains while changing things"-bugs. :rofl:

**Comment by BuckarooBanzay** [comment]
@S-S-X Is the current state (semi-) stable for testing? I'd like to test-drive this again :smile:

**Comment by S-S-X** [comment]
> @S-S-X Is the current state (semi-) stable for testing? I'd like to test-drive this again

@BuckarooBanzay now that those few brain farts have been fixed it should be stable. I did not really test node digging and placement at all, did some changes there.

I think digging and placement should work but API for those is not stable meaning that code for digging and placement will probably still see some updates and refactoring.
Currently logic is scattered between cables.lua and network.lua and only some core logic is moved (`remove_network_node`, `add_network_branch`) to network.lua but not all. There's still some parts of redundant stuff that should be moved to network.lua.

tldr. Should be good enough for testing.

**Comment by S-S-X** [comment]
Oh yet another thing that is not properly finished is network timeouts, network should be aware of timeouts instead of having special tracking pieces tied to globalstep stuff. I mean there should be single ABM for for network activation and one place where network timeout is stored and correct place is in network hash.
So these have to be changed still: Remove switch ABM from globalstep, remove timeout management from globalstep, use off_delay_seconds for network timeout.

File: technic/machines/switching_station_globalstep.lua [hunk 2]
```diff
@@ -111,34 +72,30 @@ minetest.register_globalstep(function(dtime)
...
+					network.skip = 30
 				elseif switch_diff > 150000 then
-					switch.skip = 20
+					network.skip = 20
 				elseif switch_diff > 75000 then
-					switch.skip = 10
+					network.skip = 10
 				elseif switch_diff > 50000 then
-					switch.skip = 2
+					network.skip = 2
```
[the remark this answers is not in the document]
Comment (S-S-X) [reply technic/machines/switching_station_globalstep.lua:81]: It might be possible to divide laggy/large network execution into buckets where single cycle executes only part of network, there's few problems with this btu I think everything can be solved without complicating things too much:
* Network has 20 machines, 10 first gets executed, first few machines gets removed, next cycle is not balanced anymore without careful index tracking, things gets complicated.
* Supply converters might do weird things like free energy or cutting off continuously for short periods of time.
* Network output probably has to be cached, demand will fluctuate depending on what machines got selected for partial execution. Demand is significantly lower than what it is when whole network gets executed as whole. This might not be a problem as if it gets back to full cycles it will simply stop if demand > supply.
* Would be very good if battery arrays would be cached as single huge virtual battery unit and actual batteries updated during one separate cycle, destroying and forcing rebuild of network between cycles (constructor/piston/node breaker) if timing is correct might create infinite battery capacity and cause even more lag while giving benefits for user.

Probably better to let it be like that for now and see later if there's better solutions for that, changing it now might complicate things too much.

**Comment by S-S-X** [comment]
There's still that not-so-nice digtron hack, it is not fine imo. Possibly should be disabled or that hacky patch extended to also fix that older tron power connector bug to disallow running trons without actual energy source.
I'll try to check if anything can be done through extending that dirty workaround with digtron bug patch, depends on how well digtron internal functions / api can be overridden...

**Comment by S-S-X** [comment]
Removed digtron compatibility hack (just commented out dofile) as other digtron bugs around HV connector cannot be fixed without actually fixing digtron code. I did not find much about this bug by searching around but as HV connector is not really used much I think now that bug and mechanics behind it are actually well known it is better to not allow using power connector until problem is actually fixed in digtron mod.

All that is really needed with network-ng functionaly would be checking every digtron cycle if any of locations around HV power connector is cable, then it is easy to check if cable is part of active network without searching or reading meta.

However that way is not fully compatible with upstream technic.
Also verified that this same bug appears with upstream technic master branch.

**Comment by BuckarooBanzay** [comment]
> HV connector is not really used much

The digtron HV connector has very limited use-cases, only stationary contraptions come to mind (farming, maybe bulk ship-building)

> I think now that bug and mechanics behind it are actually well known it is better to not allow using power connector until problem is actually fixed in digtron mod.

Worst that can happen is that the digtron runs a few cycles "for free" right?

That won't be the only thing buggy while using the digtron, just saying :wink:

**Comment by S-S-X** [comment]
> few cycles "for free" right?

Yes, only that count of free cycles is not actually known but I believe it can be infinite. I had only about 70m test track with power cable lasting only first 20m for tron and that it cleared fully all the way to stopper blocks at end of track. So there it was 50 free cycles.

**Comment by S-S-X** [comment]
This should be ready for merge, would be good to do some more serious testing first.
@BuckarooBanzay possible to do some testing @ pandorabox test server?

I think that even if there's still some bugs it is already a lot better and brings more good than bad but should still check if there's some new bugs that could cause crashes.

Latest changes with cable and machine placement / digging might possibly contain some bugs that can crash server even while I already tested some weird things locally.

**Comment by BuckarooBanzay** [comment]
> This should be ready for merge, would be good to do some more serious testing first.

oh, already :astonished:

>  @BuckarooBanzay possible to do some testing @ pandorabox test server?

I've updated the mod on the test-server but i don't know if i have time for testing myself today.

**Comment by S-S-X** [comment]
Every exiting network I checked so far on test server seems to work correctly. Supply converters blinking "Has no network" infotext fixed with 7fbd348, that's not yet in test server and also it is not actually affecting actual input/output of converters.

There's still these cable digging / placement bug where one has to place machine again if cable that connects to more than digged 
screenshot_20201021_022037

Also another possibly related to above:
1. cable between machine and switch but not below machine is digged
2. digged cable is placed back
3. cable below machine is digged
4. machine wont be removed from network

Machine draws power from network but without actual cable connecting machine to network. MV furnace in screenshot continues to work and draw power from battery box:
screenshot_20201021_151521

**Comment by OgelGames** [comment]
> Machine draws power from network but without actual cable connecting machine to network.

I have actually seen that bug happen once before awhile ago, never tested to see if it could be reproduced though...

**Comment by S-S-X** [comment]
> > Machine draws power from network but without actual cable connecting machine to network.
> 
> I have actually seen that bug happen once before awhile ago, never tested to see if it could be reproduced though...

At least I could not reproduce that using exact same described method. I'll try to add tests for dig/place stuff and see if I can find way to fix this, probably directly related to cable/machine on_construct/on_destruct callback changes.

**Comment by S-S-X** [comment]
I think most placement and digging bugs fixed and tests added for those.
Noticed one bug that I did not fix but added note about that to sources: Digging single cable will remove machine from network even if machine has multiple connected cables to same network, 2 ways to work around: either dig another connected cable and place it again or dig machine and place it again.

**Comment by S-S-X** [comment]
After testing is done well enough what would be nice way to merge this?
Clean up commits first (_rewrite history_), just squash all together (_maybe too much for single commit imo_), divide into few branches (_rewrite adding tests/network code/hacks through new branches_) or just merge every commit as is into master...?

**Comment by BuckarooBanzay** [comment]
> After testing is done well enough what would be nice way to merge this?

Squash if possible (if you don't mind "loosing" _your_ history)
Otherwise merge every commit as-is, there are already 1250 commits so far.

Thats a lot of work there btw :astonished:

**Comment by S-S-X** [comment]
> Squash if possible (if you don't mind "loosing" your history)

Well, yes probably better than merge commit. Probably still not very useful when single commit is +3,045 −746, some over 2k lines for tests. not that fun going through diff if someone wants to find something later but maybe that's manageable.

Maybe I can try to squash some of it but not sure if that is possible, and if it seems too complicated to clean up then just squash all together.

**Comment by S-S-X** [comment]
I think this would work okay, leaving it there for copy paste. Just add commits if some still added here in this branch. That's 44 -> 11 commits if I counted right and keeps about sane history spearating most unit test / bugfix stuff from features.
```
p 0fc0285 Move basic global network funcs to network.lua
p 8acc3d5 Add busted unit testing for technic
s 2df05b4 Fix test framework and luacheck, Add metadata cleanup ABM
s 58c439f Added busted and contentdb badges
p aaf15e7 Network refactoring, proper multi switch support WIP
s 9060d14 Update network utils, globalstep use networks, update switch ABM
f e66d3c8 De bug
s c0ed837 Removing switching stations from network
r 26d3987 Fix digiline event, luacheck warnings, tests
s ccfd796 Add fake get_modpath for busted
p 693ba64 Remove metadata for supply,demand,battery_count,battery_charge,battery_charge_max
f 9a1f598 Fix battery charge key typo
s 9991c6e Remove switching station from networks
s 0378d2c Disable few debug prints, add removed machines to LBM
s ff491da clear_networks refatoring #95
s 93cd173 Functions remove_network_node and add_network_branch #95
s ec96f7c Fix undef network_id, remove flatten #95
p ac77fa3 Tests for network constructor
s 0a9d943 Cleanup network.lua
s c840a98 Add minetest vector.lua
s 1fdd78d Add tests for network constructor
p eddc3d8 Add machines to all_nodes, no holes in arrays, check if node is added already
s c3cfc8d Comment out debug prints
s dd8fcb4 Network node place/dig, simple remove_network test
s 0d0a0bb Luacheck warnings
s 5cd5ca3 Fix network check
s 65e86aa Fix overridden global table
p 88a9b42 Network handles timeouts, remove second switch ABM
p 8912a3d #100 Add compatibility hack for digtron
p 8b37cf6 Network node placement, utilize locals for frequent names
s 14a7d1a #95 Place/remove network nodes. Cleanup sw_pos
d 42a5f00 Disable digtron meta compatibility hack
d 58d9d0f Revert "Disable digtron meta compatibility hack"
r fd7e08b #100 Timeout check back to 1 second / SC infotext / reduce timeout ABM workload
s 7fbd348 #96 SC infotext / reduce timeout ABM workload
r a74cb1a Network build/dig tests and bugfixes
s 0ae2a97 Fix few cable place/dig issues
s a8ef0ce Tests for machine build/dig
s eff3512 Fix machine build/dig bugs
s 63d9b88 Fix test for machine building
s 65affc1 Fix machines acting like cables when placed
```

**Comment by S-S-X** [comment]
@OgelGames as you mentioned that you might be able to do some testing here now would be good time to do final tests.
@BuckarooBanzay could also be good if possible to update this to pandorabox test server again after many cable placement/digging bugs were fixed (checkmark list in description).

File: technic/machines/network.lua [hunk 3]
```diff
@@ -123,10 +132,28 @@ local function touch_node(tier, pos, timeout)
 		-- this should get built up during registration
 		node_timeout[tier] = {}
 	end
-	node_timeout[tier][poshash(pos)] = timeout or 2
+	node_timeout[tier][poshash(pos)] = timeout or 5
```
Comment (S-S-X) [inline technic/machines/network.lua:135]: This was updated to somewhat mitigate problems caused by throttling, last technic problems at pandorabox (machines showing/blinking no network often) were caused because:
* Timeout ABM frequency was set back to 1sec / 100% to make digtron better
* Throttling at max lag 1.5 or more causes technic run to have at least 3 sec intervals.
* Machines will time out between globalstep because ABM timeout is 2 seconds.

So changed this to 5 seconds for now to somewhat mitigate issues when throttling is active.

Comment (S-S-X) [reply technic/machines/network.lua]: Changed again to be dynamic based on current throttling...

File: technic/machines/network.lua [hunk 4]
```diff
@@ -132,7 +137,7 @@ local function touch_node(tier, pos, timeout)
 		-- this should get built up during registration
 		node_timeout[tier] = {}
 	end
-	node_timeout[tier][poshash(pos)] = timeout or 5
+	node_timeout[tier][poshash(pos)] = timeout or default_timeout
```
Comment (S-S-X) [inline technic/machines/network.lua:140]: Changed again to be dynamic based on current throttling interval rounded up + 1.

**Comment by OgelGames** [comment]
Did some testing, and overall it's working much better, but I did find some bugs:

Moving a switching station with a piston gets rid of the network for machines connected to it, even though it is still connected, and the switching station shows the network is still working:

screenshot_20201025_170608
screenshot_20201025_170614

To reproduce:
1. Create a functional network.
2. Push or pull the switching station with a piston to a different cable that is already connected to the network.


Separate networks don't connect when adding cables that should connect them, and the switching stations don't get overloaded:

screenshot_20201025_170706
screenshot_20201025_170714

To reproduce:
1. Create two separate networks, each with a switching station. 
2. Place cables to connect the networks.


Also, not a bug but something that I think should be changed; placing multiple switching stations should not overload and stop the network, but instead the first switching station should take priority, with the additional ones being disabled.

**Comment by S-S-X** [comment]
> Push or pull the switching station with a piston to a different cable that is already connected to the network.

That is bit weird but probably there's some leftovers in switching station code, have to check that.

> placing multiple switching stations should not overload and stop the network, but instead the first switching station should take priority, with the additional ones being disabled.

That is not what happens there, you can place as many switching stations you want and nothing happens in network, instead of placing switching station you merged 2 different networks together.

**Comment by S-S-X** [comment]
> Push or pull the switching station with a piston to a different cable that is already connected to the network.

Actually I do know where this comes from, this happens because switching station is not network controller but globalstep does check if there is switching station node above network reference cable. This check causes globalstep to stop for this network and therefore machines times out, network is not updating machines anymore.

There's 2 ways to fix this: remove that check, it will allow network to run. It will also allow network to run without switching station until whole network times out. Or destroy network if switching station above reference cable disappeared, this is what happens when switching station is digged.

Switching station has always been part of the networks so maybe keep it and destroy network if globalstep sees that switching station is missing. ABM will then rebuild network at new location. Now when network is not destroyed, moved switching station keeps network active but globalstep refuses to run network because it does not see switching station there.

**Comment by OgelGames** [comment]
> > placing multiple switching stations should not overload and stop the network, but instead the first switching station should take priority, with the additional ones being disabled.
> 
> That is not what happens there, you can place as many switching stations you want and nothing happens in network, instead of placing switching station you merged 2 different networks together by placing cable.

Hmm... in the first test I did, placing multiple switching stations did overload the network, but I just tested it again and now it works... 😕

**Comment by S-S-X** [comment]
> placing multiple switching stations did overload the network

That should not be even possible unless for some reason network id was wrong, did you jump cables? That can cause it because currently caches are not at all compatible with jumpdrive, reason is that caches wont get rebuilt that easily anymore and that's mostly because switching station is not part of the network or network machine anymore.

But I do have near ready jumpdrive compatibility shim for that, network moving done. just edge network handling still not complete. That will be handled by destroying networks crossing edges at source or connecting at edges at target.

**Comment by S-S-X** [comment]
Should probably also test if pulling cable with piston rebuilds networks, @OgelGames did you test this one?

**Comment by OgelGames** [comment]
> Should probably also test if pulling cable with piston rebuilds networks, @OgelGames did you test this one?

Not yet, but I will, going to do more testing for the latest commit (413701d).

**Comment by OgelGames** [comment]
So looking at the networks using `//lua print(technic.active_networks[technic.sw_pos2network(<pos>)])` proved somewhat helpful in figuring out what was happening.

The first bug I mentioned has been fixed, where moving the switching station to a different connected cable broke the network, however there is still a bug with the infotext of the switching station not updating when it is disconnected from the network:

screenshot_20201025_200502

The second bug still exists though, where networks don't combine when connecting them with a cable: (which I confirmed by dumping the network tables)

screenshot_20201025_200143
screenshot_20201025_200152


Also moving a cable with a piston does seem to rebuild the network, the table id is different after moving, and it behaves as it should.

**Comment by OgelGames** [comment]
Looking at the code, maybe the infotext should be updated here to something like "Swiching Station has no network":
https://github.com/mt-mods/technic/blob/413701d00b53d140feac11528465f6130e8d947f/technic/machines/switching_station.lua#L172-L175

**Comment by S-S-X** [comment]
> Quickly looking at the code, maybe the infotext should be updated here to something like "Switching Station has no network":
> https://github.com/mt-mods/technic/blob/413701d00b53d140feac11528465f6130e8d947f/technic/machines/switching_station.lua#L172-L175

Yes, this would work. Even better would probably be doing infotext update a bit deeper to also change infotext for newly placed switching stations, if cable is not found infotext could be updated in start_network function where it checks for cable.

**Comment by S-S-X** [comment]
> The second bug still exists though, where networks don't combine when connecting them with a cable: (which I confirmed by dumping the network tables)

Quoting myself:
> But of course merging networks could be improved, currently it does not happen *in master branch*. However this behavior does not change with this PR so I don't want to add more here. Maybe create separate issue to improve merging networks together.

So, separate issue for that. **This is not to be fixed in this branch** as it behaves exactly like that in master and it is not simple feature to add that unless adding dirty hacks. Not good to add anything new here anymore, instead fix bugs introduced in this branch.

**Comment by S-S-X** [comment]
> Also moving a cable with a piston does seem to rebuild the network, the table id is different after moving, and it behaves as it should.

This probably should be working, might be that piston does not use on_construct/on_destruct callbacks at all. If that's the case then have to look for alternate callback to fix this.

**Comment by OgelGames** [comment]
>  it behaves exactly like that in master

Ah ok, I didn't realize that's what you meant 👍 

> might be that piston does not use on_construct/on_destruct callbacks at all.

AFAIK, those are called by the engine whenever a node is added or removed, regardless of how it was done or what did it.

**Comment by S-S-X** [comment]
> AFAIK, those are called by the engine whenever a node is added or removed

Yes and it seems to do that as networks are divided into two when pulling cable out with piston.
I just remembered that on_construct / on_destruct was bit different and those are but depends what functions are used to place node. Almost all cases those will be called. Docs say:
> -- Not called for bulk node placement (i.e. schematics and VoxelManip).

Problem is cable placement does not work correctly, only removing works correctly. Added tests already.

**Comment by BuckarooBanzay** [comment]
>  @BuckarooBanzay could also be good if possible to update this to pandorabox test server again after many cable placement/digging bugs were fixed (checkmark list in description).

@S-S-X a bit late to the party, sorry: `network-ng` is updated to https://github.com/mt-mods/technic/pull/96/commits/b24aa0db083688ba0ab01507e94f8faa99f4ac3b on the test-server, ping me up on discord/irc if there are issues :+1:

**Comment by S-S-X** [comment]
> > @BuckarooBanzay could also be good if possible to update this to pandorabox test server again after many cable placement/digging bugs were fixed (checkmark list in description).
> 
> @S-S-X a bit late to the party, sorry: `network-ng` is updated to [b24aa0d](https://github.com/mt-mods/technic/commit/b24aa0db083688ba0ab01507e94f8faa99f4ac3b) on the test-server, ping me up on discord/irc if there are issues

Actually that was pretty good timing as OgelGames also found few things to fix in between :smile:

**Comment by S-S-X** [comment]
Checked test server a bit and noticed that some supply converters not working at all, have to test those better. There might be some problem with updating input or output side depending on how those are connected.
Some did seem to work just fine without problem but some did not work at all.

**Comment by BuckarooBanzay** [comment]
>  Also lost connection so was this crash or something else @BuckarooBanzay

There were several crashes:
```
-rw-r--r-- 1 root root    0 Oct 25 20:15 crash_2020-10-25_19-15.txt
-rw-r--r-- 1 root root  125 Oct 25 21:58 crash_2020-10-25_20-58.txt
-rw-r--r-- 1 root root    0 Oct 26 07:36 crash_2020-10-26_06-36.txt
```

The one with 125 bytes contains:
```
2020-10-25 20:55:21: ERROR[Server]: suspiciously large amount of objects detected: 71 in (465,0,1451); removing all of them.
```

This is still the old segfault issue which seems more prominent on the test-server (possibly a clue? :shrug:)

**Comment by OgelGames** [comment]
> some supply converters not working at all, have to test those better. There might be some problem with updating input or output side depending on how those are connected.

Well I tested them a bit, and I think you are right. When placing the converter last (after cables and switching stations), it would not work until both networks were updated by something else (replacing a cable or SS), but if converter was placed before either the cables or switching stations, then it would work.

**Comment by OgelGames** [comment]
Seems like the supply converters are mostly fixed, but I found that they sometimes won't connect to the network when there are multiple conversions:

screenshot_20201029_223945

The first SC placed almost always connects, but the second one doesn't until cables/SS are replaced. This doesn't seem to effect SC placed in parallel though (connecting the same networks).

**Comment by S-S-X** [comment]
SC inserts duplicates to RE_nodes when cable connecting to network is placed nearby, this should be checked and SC added to RE_nodes only if PR_nodes was added. Removing SC leaves duplicates, only last entry is removed from tables (which is correct behavior if everything else follows spec).

When placed SC still does not attach to networks correctly, Actually when placed between clean networks it does not attach to anything. Rebuilding network cache completely will also attach SC correctly, this is what makes some existing networks work better with SC placement but I'm pretty sure it works only because there's another bug with cleanup.

**Comment by S-S-X** [comment]
SC and network cleanup fixed, better support for multi tier machines added.

**Comment by S-S-X** [comment]
Cleaned up and merging to master, users test rest best in production right
